### PR TITLE
Add iterative inertia tensor calculation

### PR DESCRIPTION
**Pull request by robjmcgibbon**
This PR consolidates the reduced and non-reduced inertia tensor calculations, and introduces an iterative method for calculating the tensors (see https://arxiv.org/abs/2404.01368). I've also removed inertia tensor calculations from aperture properties, because too large a fraction of the output files were being dedicated to storing this information. 

There was a small bug in the old version of `get_projected_inertia_tensor` which meant that the units were off by a factor of `a^2`, but otherwise the new `InertiaTensorNoniterative` and `InertiaTensorReducedNoniterative` agree with the old methods.

We could still probably save less inertia tensors, but that's easy to edit in the parameter file.

### CODE REVIEW

**Comment by robjmcgibbon** [comment]
I've added an assert to check that properties which should be dimensionless actually are. The DopplerB parameter was also incorrect by a factor of `a`, which I have fixed

**Comment by robjmcgibbon** [comment]
For me TODO: update units depending on how https://github.com/SWIFTSIM/SOAP/pull/75 is finalised

File: SO_properties.py [hunk 1]
```diff
@@ -565,16 +564,43 @@ def spin_parameter(self) -> unyt.unyt_quantity:
     def TotalInertiaTensor(self) -> unyt.unyt_array:
         """
         Inertia tensor of the total mass distribution.
+        Computed iteratively using an ellipsoid with volume equal to that of
+        a sphere with radius SORadius. Only considers particles within the SORadius
```
Comment (VictorForouhar) [inline SO_properties.py:568]: Is the volume of the ellipsoid well-within the equivalent SO sphere? In other words, would we pick up particles on the edges of a heavily stretched ellipse that are beyond the SO radius?

File: SO_properties.py [hunk 2]
```diff
@@ -565,16 +564,43 @@ def spin_parameter(self) -> unyt.unyt_quantity:
...
+        """
+        if self.Mtotpart == 0:
+            return None
+        return get_inertia_tensor(self.mass, self.position, self.SO_r, reduced=True)
+
+    @lazy_property
+    def TotalInertiaTensorNoniterative(self) -> unyt.unyt_array:
+        """
+        Inertia tensor of the total mass distribution.
+        Computed in a single iteration using particles within the SORadius.
```
Comment (VictorForouhar) [inline SO_properties.py:589]: Might be clearer if you add that one iteration is equivalent to using particles within the SO sphere, without deforming the volume into an ellipsoid. Some people may assume this step is iteration 0 (i.e. you do no update the shape of volume), rather than iteration 1.

File: aperture_properties.py [hunk 3]
```diff
@@ -144,8 +144,6 @@ def MetalFracStar(self):
     get_angular_momentum,
     get_angular_momentum_and_kappa_corot,
     get_vmax,
-    get_inertia_tensor,
-    get_reduced_inertia_tensor,
```
Comment (VictorForouhar) [inline aperture_properties.py:148]: This is removed because we will not compute the inertia tensor for spherical apertures, right?

File: kinematic_properties.py [hunk 4]
```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
...
-def get_projected_inertia_tensor(
-    mass: unyt.unyt_array, position: unyt.unyt_array, axis: int
-) -> unyt.unyt_array:
-    """
+    '''
+    # Shouldn't calculate an inertia tensor for small numbers of particles
+    if position.shape[0] < 20:
+        return None
+
+    # Remove particles at centre if calculating reduced tensor
```
Comment (VictorForouhar) [inline kinematic_properties.py:256]: Why is this needed? Will it blow up the value otherwise?

File: kinematic_properties.py [hunk 5]
```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
...
+    eig_vec = np.array([
+        [1, 0, 0],
+        [0, 1, 0],
+        [0, 0, 1],
+    ])
+
+    for _ in range(max_iterations):
+        # Calculate shape
+        old_q = q
+        q = np.sqrt(eig_val[1] / eig_val[2])
```
Comment (VictorForouhar) [inline kinematic_properties.py:283]: Why is this ratio better to use instead of `eig_val[0] / eig_val[2]`?

File: kinematic_properties.py [hunk 6]
```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
...
-
-    return Itensor
-
-
-def get_projected_inertia_tensor(
-    mass: unyt.unyt_array, position: unyt.unyt_array, axis: int
-) -> unyt.unyt_array:
-    """
+    '''
+    # Shouldn't calculate an inertia tensor for small numbers of particles
```
Comment (VictorForouhar) [inline kinematic_properties.py:252]: This condition exists to handle cases where the halo may have more than 100 particles but less than 20 within a SO radius, right? If so, mention this to make a distinction from the 100 particle limit imposed at the subhalo level.

File: kinematic_properties.py [hunk 7]
```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
...
+        q = np.sqrt(eig_val[1] / eig_val[2])
+        s = np.sqrt(eig_val[0] / eig_val[2])
+        p = np.sqrt(eig_val[0] / eig_val[1])
+
+        # Break if converged
+        if abs((old_q - q) / q) < tol:
+            break
+
+        # Calculate ellipsoid, determine which particles are inside
+        axis = R * np.array([
```
Comment (VictorForouhar) [inline kinematic_properties.py:292]: Do you remember where you got these formulaes from? I would like to double check the expressions.

File: kinematic_properties.py [hunk 8]
```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
...
+            tensor /= norm[:, None, None]
+        tensor = tensor.sum(axis=0)
+        eig_val, eig_vec = np.linalg.eigh(tensor.value)
+
+    return np.concatenate([np.diag(tensor), tensor[np.triu_indices(3, 1)]])
+
+
+def get_projected_inertia_tensor(mass, position, axis, radius, reduced=False, max_iterations=20):
+    '''
     Get the inertia tensor of the given particle distribution, projected along the
```
Comment (VictorForouhar) [inline kinematic_properties.py:314]: This can lead to confusion. As we discussed the other day, e.g. is it the inertia tensor of the projected particle distribution, or the projected inertia tensor of the particle distribution?  From our conversion, I think it is the former, but the comment makes me think of the latter.

File: property_table.py [hunk 9]
```diff
@@ -354,44 +354,45 @@ class PropertyTable:
...
-        "BaryonInertiaTensor": (
-            "BaryonInertiaTensor",
+        "DarkMatterInertiaTensor": (
+            "DarkMatterInertiaTensor",
             6,
             np.float32,
             "kpc**2",
-            "3D baryon inertia tensor computed from the baryon (gas and stars) mass distribution, relative to the centre of potential. Stores diagonal components and one off-diagonal triangle as (1,1), (2,2), (3,3), (1,2), (1,3), (2,3).",
-            "baryon",
+            "3D inertia tensor computed iteratively from the DM mass distribution, relative to the halo centre. Diagonal components and one off-diagonal triangle as (1,1), (2,2), (3,3), (1,2), (1,3), (2,3).",
```
Comment (VictorForouhar) [inline property_table.py:362]: Would it be useful to indicate somewhere how to recover the inertia tensor from this subset of the flattened representation?

File: property_table.py [hunk 10]
```diff
@@ -354,44 +354,45 @@ class PropertyTable:
             False,
             ["PartType5/SubgridMasses", "PartType5/Velocities"],
         ),
-        "BaryonInertiaTensor": (
-            "BaryonInertiaTensor",
+        "DarkMatterInertiaTensor": (
```
Comment (VictorForouhar) [inline property_table.py:357]: Will people clearly know that the "Inertia Tensor" is not the usual (non-iterative) "Inertia Tensor" computed by them?

File: property_table.py [hunk 11]
```diff
@@ -1673,6 +1748,17 @@ class PropertyTable:
                 "PartType5/DynamicalMasses",
             ],
         ),
+        "TotalSNIaRate": (
```
Comment (VictorForouhar) [inline property_table.py:1751]: Is this piece of code relevant to this PR?

File: property_table.py
Comment (VictorForouhar) [inline property_table.py:1]: Are we happy to not be computing projected inertia tensors of the total matter distribution? This may be relevant to gravitational lensing.

File: parameter_files/COLIBRE.yml [hunk 12]
```diff
@@ -119,6 +119,22 @@ ApertureProperties:
     NumberOfGasParticles: true
     NumberOfStarParticles: true
     SpinParameter: true
+    TotalInertiaTensor: true
+    GasInertiaTensor: true
+    DarkMatterInertiaTensor: true
+    StellarInertiaTensor: true
+    TotalInertiaTensorReduced: false
```
Comment (VictorForouhar) [inline parameter_files/COLIBRE.yml:126]: Why are we not computing some of these inertia tensor quantities?

File: parameter_files/FLAMINGO.yml [hunk 13]
```diff
@@ -325,21 +315,30 @@ SOProperties:
...
+    StellarInertiaTensorReduced: true
+    TotalInertiaTensorNoniterative: false
+    GasInertiaTensorNoniterative: false
+    DarkMatterInertiaTensorNoniterative: false
+    StellarInertiaTensorNoniterative: false
+    TotalInertiaTensorReducedNoniterative: false
+    GasInertiaTensorReducedNoniterative: false
+    DarkMatterInertiaTensorReducedNoniterative: false
+    StellarInertiaTensorReducedNoniterative: false
+    ThermalEnergyGas: true
```
Comment (VictorForouhar) [inline parameter_files/FLAMINGO.yml:341]: Is `ThermalEnergyGas` relevant to this PR?

File: parameter_files/FLAMINGO.yml [hunk 14]
```diff
@@ -418,7 +414,6 @@ SubhaloProperties:
     HalfMassRadiusGas: true
     HalfMassRadiusStars: true
     HalfMassRadiusTotal: true
-    EncloseRadius: true
```
Comment (VictorForouhar) [inline parameter_files/FLAMINGO.yml:421]: Will this be saved by default? It is useful to have

File: projected_aperture_properties.py [hunk 15]
```diff
@@ -664,40 +687,44 @@ def proj_veldisp_star(self) -> unyt.unyt_quantity:
...
         if self.Mstar == 0:
             return None
-        return get_projected_inertia_tensor(
-            self.proj_mass_star, self.proj_pos_star, self.iproj
-        )
+        return get_projected_inertia_tensor(self.proj_mass_star, self.proj_pos_star, self.iproj, self.aperture_radius)
 
     @lazy_property
-    def ReducedProjectedStellarInertiaTensor(self):
+    def ProjectedStellarInertiaTensorReduced(self):
```
Comment (VictorForouhar) [inline projected_aperture_properties.py:699]: Several of the new properties are missing `-> unyt.unyt_array:`

**Review by VictorForouhar — COMMENTED** [review]
Check my comments.

File: aperture_properties.py (hunk 3 above)
Comment (robjmcgibbon) [reply aperture_properties.py:148]: Yep

File: kinematic_properties.py (hunk 5 above)
Comment (robjmcgibbon) [reply kinematic_properties.py:283]: Do you mean why do I use this as the stopping criteria?

File: parameter_files/COLIBRE.yml (hunk 12 above)
Comment (robjmcgibbon) [reply parameter_files/COLIBRE.yml:126]: We don't want to compute them for most snapshots, but they are going to be used for z=0 to compare the methods

File: parameter_files/FLAMINGO.yml (hunk 13 above)
Comment (robjmcgibbon) [reply parameter_files/FLAMINGO.yml:341]: This just got moved when I was grouping all the tensor properties together, but it was set to true before anyway

File: kinematic_properties.py (hunk 5 above)
Comment (VictorForouhar) [reply kinematic_properties.py:283]: Yes, as opposed to the other eigenvalue ratio you could take. Is the current choice more stable, or is it just a choice?

File: parameter_files/FLAMINGO.yml (hunk 14 above)
Comment (robjmcgibbon) [reply parameter_files/FLAMINGO.yml]: Good catch, this wasn't intentional. I guess it got removed when I pulled in the latest version of master branch. Added back in https://github.com/SWIFTSIM/SOAP/pull/82/commits/5107c1f26880497e774c890bb0e06edaf4838b52

File: property_table.py
Comment (robjmcgibbon) [reply property_table.py:1]: That was Joop's choice, but I can double-check with him

File: kinematic_properties.py (hunk 5 above)
Comment (robjmcgibbon) [reply kinematic_properties.py:283]: It's an arbitrary choice

File: kinematic_properties.py (hunk 7 above)
Comment (robjmcgibbon) [reply kinematic_properties.py:292]: I derived them. The volume is proportional to the product of the three axis lengths. Any axis length is proportional to the other axis lengths based on the axis ratios which we have from the eigenvalues. I get these expressions by equating the volume of the initial sphere with the volume as a function of a single axis length.

File: kinematic_properties.py (hunk 4 above)
Comment (robjmcgibbon) [reply kinematic_properties.py:256]: Yes https://github.com/SWIFTSIM/SOAP/blob/5107c1f26880497e774c890bb0e06edaf4838b52/kinematic_properties.py#L398 won't be happy

File: property_table.py
Comment (VictorForouhar) [reply property_table.py:1]: We can discuss it during tomorrow's meeting.

File: parameter_files/FLAMINGO.yml (hunk 14 above)
Comment (VictorForouhar) [reply parameter_files/FLAMINGO.yml]: Dank je.

File: kinematic_properties.py (hunk 6 above)
Comment (robjmcgibbon) [reply kinematic_properties.py:252]: I've updated the particle description in https://github.com/SWIFTSIM/SOAP/pull/82/commits/25bd4fc9b8998e9e51c14f80fda40dcb076f8ea7

File: SO_properties.py (hunk 1 above)
Comment (robjmcgibbon) [reply SO_properties.py:568]: No we won't pick up particles outside the SO radius, and the ellipsoid is basically guaranteed to stretch outside since the initial radius is equal to the SO radius. The same happens for the projected apertures.

This is something I had considered, but there's no nice solution without more major changes to SOAP.

File: property_table.py (hunk 10 above)
Comment (robjmcgibbon) [reply property_table.py:357]: I think the description makes this clear, so I'm inclined to leave the name as it is

File: projected_aperture_properties.py (hunk 15 above)
Comment (robjmcgibbon) [reply projected_aperture_properties.py]: Added in https://github.com/SWIFTSIM/SOAP/pull/82/commits/394ddd06eda98f6878b9906f5fb68193525b7c8a

File: kinematic_properties.py (hunk 8 above)
Comment (robjmcgibbon) [reply kinematic_properties.py]: Updated in https://github.com/SWIFTSIM/SOAP/pull/82/commits/394ddd06eda98f6878b9906f5fb68193525b7c8a to (hopefully) clarify that it's the former

File: SO_properties.py (hunk 1 above)
Comment (VictorForouhar) [reply SO_properties.py:568]: Leaving for future reference: In the meeting we agreed to explore the possibility of computing shapes for apertures when we have loaded particles at larger apertures (e.g. R200 shapes when we have particles within R100).

File: property_table.py
Comment (VictorForouhar) [reply property_table.py:1]: In the meeting, we agreed this should be added.

Comment (robjmcgibbon) [reply property_table.py:1]: Done in https://github.com/SWIFTSIM/SOAP/pull/82/commits/1cfeeb4049ed647bf4d3485594c8ae105c86be6c

**Comment by robjmcgibbon** [comment]
I am now handling cases where we need to use particles which are outside the initial sphere/circle. For subhalo properties this wasn't a problem since we are already passing all bound particles. For projected apertures I now pass all bound particles, rather than only the bound particles within the aperture. For SO_properties I pass all particles which have been loaded. If the ellipsoid reaches outside the search radius then I raise an error.

I have renamed ReadRadiusTooSmallError to SearchRadiusTooSmallError

File: SO_properties.py (hunk 1 above)
Comment (robjmcgibbon) [reply SO_properties.py]: https://github.com/SWIFTSIM/SOAP/pull/82/commits/c0c1d4ce339c6f083e13c9d8b35b094648e26ae6

**Comment by robjmcgibbon** [comment]
Closing since we needed to merge to a different branch in https://github.com/SWIFTSIM/SOAP/pull/91
